### DISCUSSION TITLE
What would it take to interact with Cue in Ruby?

### DISCUSSION THREAD

**Discussion by jer-k** [cue-lang/cue#1190] [Q&A]
Okay a bit of a big hairy question here, but I thought I would ask as I've been exploring this. Also I'm quite new to Cue so if I get some terminology wrong, I apologize in advance.

Super high level info: 

I'm working on a building out a new schema, our old one is in YAML and is not very flexible. I came across Cue, presented to the team, and everyone agreed that it looked like the solution we want to use. Our schema can be thought of as very similar to Github Actions. There is a repository that started an implementation of Github Actions schema in Cue at https://github.com/hofstadter-io/ghacue/blob/main/schema.cue if you are unfamiliar. 

Our team will define the base schema and provide users with a starting point (which I'm going to refer to as a template), but we will encourage users to make their own changes to these templates.

```
#OurTemplate: {
  name: string
  branch: string
}
```
and then the template for a user would be

```
UserTemplate: #OurTemplate & {
  name: "UserTemplate",
  branch: "main"
}
```
To generate this `UserTemplate` I came across using `--inject` on the CLI and using `@tag` so I could do `cue eval example.cue --inject name=UserTemplate --inject branch=main` and achieve the desired result. This is pretty contrived example and I believe that the schema we're working to build will be a bit more complex, but we may be able to make it work using the CLI.

Now in the Ruby side:
Since our application is running in a Docker container, I can install Cue into the container, and then shell out of Ruby into the container access Cue via the CLI. Doing so will allow me to use `--inject` approach. This is okay and will likely allow me to get started on this project.

However I was reading the docs and came across [Modify Cue Values](https://cuelang.org/docs/integrations/go/#modify-cue-values), which shows that in Go we're able to directly access the keys. Being able to do the replacement this way would be far more preferable, which leads me to the question of what would it take to interact with Cue in Ruby in a similar fashion to the way Go does it? How much code (I really had no idea where to look) is this interoperability in Go? 

Thank you for any input. I'm just exploring this idea from the standpoint of having a large Rails application and how we can integrate with Cue.

**Comment by verdverm** [comment]
You'll want to compile cue as a shared library and then use Ruby's FFI to access it. Googling for "golang ruby FFI" has results but they seem a bit out of date. If I recall correctly, Go's cgo interface has changed since then.

**Reply by jer-k** [reply]
Got it! Thanks for the direction. I'll check this out and report back.
